uiautomationutilities/src/main/java/com/microsoft/identity/client/ui/automation/broker/BrokerCompanyPortal.java

```diff
@@ -53,40 +68,225 @@ public BrokerCompanyPortal() {
+        final int randomPin = random.nextInt(10000);
+
+        final UiObject pinField = UiAutomatorUtils.obtainUiObjectWithResourceId(
+                "com.microsoft.windowsintune.companyportal:id/pin_entry_passcodeEditView"
+        );
 
         try {
-            accountSelected.waitForExists(FIND_UI_ELEMENT_TIMEOUT);
-            accountSelected.click();
-        } catch (UiObjectNotFoundException e) {
+            pinField.setText(String.valueOf(randomPin));
+        } catch (final UiObjectNotFoundException e) {
+            fail(e.getMessage());
+        }
+
+        device.pressEnter();
+
+        // confirm PIN
+
+        final UiObject pinConfirmField = UiAutomatorUtils.obtainUiObjectWithResourceId(
+                "com.microsoft.windowsintune.companyportal:id/pin_entry_passcodeEditView"
+        );
+
+        try {
+            pinConfirmField.setText(String.valueOf(randomPin));
+        } catch (final UiObjectNotFoundException e) {
+            fail(e.getMessage());
+        }
+
+        device.pressEnter();
+    }
+
+    private void openDevicesTab() {
+        // launch CP
+        launch();
+
+        try {
+            final UiDevice device = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation());
+
+            // Click Devices Tab
+            final UiObject devicesTab = device.findObject(new UiSelector().description(
+                    "Devices, Tab, 2 of 3"
+            ).clickable(true));
+
+            devicesTab.waitForExists(FIND_UI_ELEMENT_TIMEOUT);
+
+            devicesTab.click();
+        } catch (final UiObjectNotFoundException e) {
             Assert.fail(e.getMessage());
         }
     }
 
+    /**
+     * Removes a device from Company Portal (from the devices listed in CP Devices Tab)
+     */
+    public void removeDevice() {
+        // if enrollment failed, then Devices Tab is automatically opened for us
+        if (enrollmentPerformedSuccessfully) {
+            openDevicesTab();
+        }
+
+        try {
+            final UiDevice uiDevice = UiDevice.getInstance(InstrumentationRegistry.getInstrumentation());
+
+            // If Enrollment failed, then the first device on the list is in corrupted state and
+            // cannot even be removed, we need to remove the second one in the list
+            final UiObject deviceToRemove = uiDevice.findObject(new UiSelector()
+                    .resourceId("com.microsoft.windowsintune.companyportal:id/device_list_item")
+                    .index(enrollmentPerformedSuccessfully ? 0 : 1)
+            );
+
+            deviceToRemove.waitForExists(FIND_UI_ELEMENT_TIMEOUT);
+
+            // click on the device to be removed
+            deviceToRemove.click();
+
+            // Click more options in the top right
+            final UiObject threeDots = uiDevice.findObject(new UiSelector().descriptionContains(
+                    "More options"
+            ));
+
+            threeDots.waitForExists(FIND_UI_ELEMENT_TIMEOUT);
```

Comment:
Question: If these elements can't be found (due to app update, for example), do these fail in a way that's easy to understand/correct?

Reply:
Or..... should we assert the existence of these buttons? 🤔

Reply:
The test would fail with an `AssertionError` and there would be precise error logs indicating that an element with this specific id is not found. The test wouldn't know however, if an element wasn't found due to the app changing their UI or if this was simply an issue with UI Automator. IMO, the error message about the element should be enough to guide us to do the following:

- Retry Test via Automation
- If still fails, it is probably that the App changed their UI, then we should attempt manual